Add endpoint for getting adventure information

whonut opened the pull request:
This addresses #19. The URL patterns are probably not what we want but I thought that was something best discussed. Any and all comments and criticisms are welcome.

whonut committed:
Add endpoint for retrieving adventure information

Check the validity of the adventure ID in the view rather than in the
URL dispatcher so that the proper BAD REQUEST response can be returned.

whonut committed:
Fix URL routing in adventures

whonut committed:
Fix typo

whonut committed:
Name adventure_by_id URL pattern

whonut committed:
Add tests for adventure_by_id endpoint

probably-not-a-cat commented:
I'm kind annoyed that those Django api tutorials just suggest to make serialized classes, when this is clearly better! 

Just an fyi, one of your tests is failing, but not failing travis. Which also explains why codecov isn't 100%.

whonut commented:
Which test? They all pass on my machine...

probably-not-a-cat commented:
https://travis-ci.org/AdventureLookup/adventurelookup-backend/jobs/134558640#L528

probably-not-a-cat commented:
@whonut I would suggest merging with master since bad tests should fail the travis build now.

Also looks like it can't connect to the website on travis.  That'll be fun to debug ;)

pejter commented:
You should also add newline at the end of each file to comply to PEP8.

whonut committed:
Add newlines at EOF

whonut committed:
Fix test_adventure_by_id_get_success

whonut commented:
@pejter Sublime Text displays a spurious blank line when you do that apparently! That'll explain why my linter was shouting at me for leaving (what I saw as) 2 blank lines at the end of a file. I've fiddled with my preferences, won't be problem now.

I'm feeling around blind trying to fix that failing test, everything works on my machine. Sorry if I do/did something stupid.

OK that 'fix' was (as I suspected) not one. I'm out of ideas.

probably-not-a-cat commented:
@whonut I would recommend making a new branch when it comes to testing travis. That way you can keep this PR clean. You can rebase and merge back here later.

pejter commented:
`objects.create` only returns the instance of the object. You need to call `.save()`on the instance for the change to be applied to the database.
From [Django docs](https://docs.djangoproject.com/en/1.9/ref/models/instances/#django.db.models.Model): `Note that instantiating a model in no way touches your database; for that, you need to save().`

Some other points:
- You should test the JSON you get vs the actual object and not the input data.
- Also creating objects should be done in the setUp of the tests and not in the actual test(except for object creation tests).
- [Writing tests in Django](https://docs.djangoproject.com/en/1.9/topics/testing/overview/#writing-tests)
- [Testing views with a live test client](https://docs.djangoproject.com/en/1.9/topics/testing/tools/#liveservertestcase)

whonut commented:
@probably-not-a-cat OK, sorry about that. Should I squash that last one?
@pejter `create` saves it though...? [Docs](https://docs.djangoproject.com/en/1.9/ref/models/querysets/#create):

> A convenience method for creating an object and saving it all in one step. Thus:
> 
> p = Person.objects.create(first_name="Bruce", last_name="Springsteen")
> 
> and:
> 
> p = Person(first_name="Bruce", last_name="Springsteen")
> p.save(force_insert=True)
> 
> are equivalent.

Thanks for the help with the tests. I think I've done it right now but I'll wait to hear back before pushing.

I can only apologise for my ineptitude with this. Thank you both.

probably-not-a-cat commented:
I wouldn't. I tried squashing a PR before and github doesn't handle it nicely. It's not a big deal to leave it.

And no worries! Thanks for doing this!

pejter commented on web/adventurelookup/urls.py:
```diff
@@ -13,12 +13,13 @@
 
 urlpatterns = [
     url(r'^admin/', admin.site.urls),
+    url(r'^adventures/', include('adventures.urls'))
```
Views should be namespaced to avoid name collision. You can do that by specifying the `namespace` parameter to include:
`include('adventures.urls', namespace='adventures')`

pejter commented on web/adventures/test_adventures.py:
```diff
@@ -57,3 +58,33 @@ def test_URLListField_get_prep_value(self):
+        self.client = Client()
+
+    def test_adventure_by_id_get_success(self):
+        test_adv = Adventure.objects.create(name="LMoP", links=["www.google.com", "another.website.io"])
```
You can initialize the model with a dictionary in which keys match field names with:
`new_mymodel = MyModel.objects.create(**data_dict)`
This would ensure that the model is properly initialized with your correct input and make it DRY.

pejter commented on web/adventures/test_adventures.py:
```diff
@@ -57,3 +58,33 @@ def test_URLListField_get_prep_value(self):
+        self.assertEqual(actual_data, correct_data)
+
+    def test_adventure_by_id_get_404(self):
+        notfound = self.client.get('/adventures/adventure/15')
```
Instead of hard-coding urls you should use [`urlresolvers.reverse(viewname)`](https://docs.djangoproject.com/en/1.9/ref/urlresolvers/#reverse) with `args` or `kwargs` to avoid errors when URLs change.

pejter commented on web/adventures/views.py:
```diff
@@ -0,0 +1,21 @@
+
+    # Adventure IDs must be integers, respond with BAD REQUEST if an
+    # invalid ID is given.
+    is_valid_id = all(char.isdigit() for char in adventure_id)
```
This probably should be handled with a [URL regex](https://docs.djangoproject.com/en/1.9/topics/http/urls/#example) unless we have a specific reason to return 400 instead of 404.

whonut replied:
My thinking was that it gives more specific information about the error to the caller than a 404 that could mean multiple things. Would that be reason enough? I agree it's nicer to handle with the regex, I was originally going to do it that way.

pejter replied:
This won't be a public facing API and I think if we'll ever want more verbosity we can always make the change. For now I think the regex would be easier to understand and the natural place for this kind of validation.

whonut committed:
Namespace 'adventures/' URL patterns

Will allow for the use of urlresolvers.reverse.

whonut committed:
Use setUpTestData instead of setUp

Creation of test data should be done in a setup function, not in the
test itself. setUpTestData does this once for the entire class of
tests, where setUp does it before each.

whonut committed:
Use reverse instead if hard-coded URLs

whonut committed:
Validate adventure ID in URL regex, not in view

The API will not be public facing, so the added information from
status code 400 vs 404 is not particularly necessary. The URL
regex is a more natural place for validation.

Also removes now-unnecessary test for status code 400.

whonut committed:
Test returned JSON against the database object

whonut committed:
Initialise test adventure from dict

probably-not-a-cat commented:
Just an FYI, sometimes travis builds have issues that are not because of your code, which happens to be why it failed here. I would just wait a bit and restart the build.

whonut committed:
Make adventure_by_id into a class-based view

The switch to class-based rather than function-based views is motivated
by the future inclusion of a generic JSONResponse view, from which this
and other specific views would inherit.

Currently only get requests are handled specially, but the addition to
dispatch means that other HTTP verb methods have a consistent way to
access the adventure ID.

whonut commented:
Just FYI, I've made a [class-based version](https://github.com/whonut/adventurelookup-backend/commit/1ca9bc858871be02dcd80291c705051b02333afa), if that's what we're going with.

probably-not-a-cat commented:
This looks good to me. 👍

whonut committed:
Have AdventureById.get accept only one argument

The adventure_id URL parameter is handled by the dispatch method, so
get only needs to accept a HttpRequest, not arbitrary args and kwargs.
As a result, super().dispatch should only be called with a single
HttpRequest as an argument, so that it passes only this to get.

whonut committed:
Store adventure object in attribute, not ID

Storing the adventure object associated with a particular request to
AdventureById in an instance attribute, rather than storing the unique
ID of the adventure, removes the need to get the object from the ID
separately in each method.

probably-not-a-cat commented:
Since there's no activity on here, and all concerns have been addressed, I would :shipit: (merge).

pejter commented:
Well we did settle on using the class-based approach so I was waiting for that. @whonut has to merge that into this branch and we can merge it.

whonut commented:
And I was waiting for the go-ahead 😝 Will merge.

whonut commented:
I've merged the class-based branch now. It shows up in the commits tab, if not here.
